### ISSUE TITLE
RPC 401 Unauthorized After Upgrade to version 1000950

### ISSUE THREAD

**Issue by ethought** [zcash/zcash#2444]
A couple of days ago our older Zcash client (1000650) crashed and upon restart I noticed the following error:

"errors": "Your client is out of date and vulnerable to denial of service. Please update to the most recent version of Zcash (1.0.8-1 or later). More info at: https://z.cash/support/security/"

So we upgraded our exchange to a new version of Zcash:

  "version": 1000950,
  "protocolversion": 170002,
  "walletversion": 60000,

However it seems it is now not possible to connect via regular RPC calls anymore.

Nothing else has changed in our setup. We have the same configuration file etc but only receive the following error now:

401 Unauthorized

Any suggestions would be appreciated as we have customers waiting.

**Comment by str4d** [comment]
We didn't make any authentication changes in the last few releases - the cookie-based RPC auth was added in 10.5 whereas you were running 1.0.6. But there *were* general RPC improvements from upstream merged into 1.0.7...

Is your `zcash-cli` on the same machine as your `zcashd`? If so, could you please try commenting out all configuration options beginning with `rpc` in your configuration file (ie. add a `#` in front)? That should result in it using cookie-based authentication with all-default settings. If that works (or if your `zcash-cli` is on a different machine), we would need to know which RPC configuration options you are using.

**Comment by arcalinea** [comment]
This is high priority for user support, please triage appropriately in work queue

**Comment by ioptio** [comment]
@ethought ping. Can you please answer above questions so we can address this appropriately?

**Comment by Vectorist** [comment]
I reverted to v1.0.6 and Perl scripts could call via rpc, but cannot on the current version.  I have to fix my python environment variables, and then I can test out python 3.6 on the current versioni of Zcash.  Zcash-cli rpc's work on 1.0.9. I am on ubuntu 16 LTS. Let me know if would like me to test anything else.

**Comment by daira** [comment]
Can you give a transcript of a full HTTP request sent by your script including headers, and the responses from 1.0.6 and from a more recent version that didn't work?

**Comment by Vectorist** [comment]
I have yet to certify as a SysAdmin, so I hope this covers everything:

   use JSON::RPC::Client;
...
  $client->ua->credentials(
     #'10.0.0.2:8232', 'jsonrpc', 'Daedalus' => 'redacted'  # REPLACE WITH YOUR bitcoin.conf rpcuser/rpcpassword
     'localhost:8232', 'jsonrpc', 'Daedalus' => 'redacted'  # REPLACE WITH YOUR bitcoin.conf rpcuser/rpcpassword
      );

This is just a copy paste script from bitcoinwiki that I use to troubleshoot electrum-server.  The recent version throws an 401 Unathorized error.
[log.txt](https://github.com/zcash/zcash/files/1134068/log.txt)
[log1.txt](https://github.com/zcash/zcash/files/1134069/log1.txt)

**Comment by str4d** [comment]
The script mentioned above is [here](https://en.bitcoin.it/wiki/API_reference_(JSON-RPC)#Perl). I modified it to work with Zcash on Ubuntu Xenial:

```perl
use JSON::RPC::Legacy::Client;
use Data::Dumper;

my $client = new JSON::RPC::Legacy::Client;

$client->ua->credentials(
   'localhost:8232', 'jsonrpc', 'user' => 'password'  # REPLACE WITH YOUR zcash.conf rpcuser/rpcpassword
    );

my $uri = 'http://localhost:8232/';
my $obj = {
    method  => 'getinfo',
    params  => [],
 };

my $res = $client->call( $uri, $obj );

if ($res){
    if ($res->is_error) { print "Error : ", $res->error_message; }
    else { print Dumper($res->result); }
} else {
    print $client->status_line;
}
```

With this test script, I am able to reproduce both the `401 Unauthorized` error on both master, and the success on 1.0.6. Here is a version bisection:

- 1.0.6: Success
- 1.0.7: Success
- 1.0.8-1: 401 Unauthorized
- master: 401 Unauthorized

So this is definitely unrelated to the RPC changes we pulled in from upstream in #2100, but *could* be related to the libevent changes we pulled in from upstream in #2176.

Note that the test script uses what is now a legacy API, so it's possible that this has a known cause, and that the Bitcoin wiki should be updated.

**Comment by str4d** [comment]
Found the bug upstream! https://github.com/bitcoin/bitcoin/issues/7462

And the fix: https://github.com/bitcoin/bitcoin/pull/7472

**Comment by ethought** [comment]
Looks like this issue is fixed. If so which version do we need to install which includes the fix?

**Comment by daira** [comment]
1.0.11-rc1 or 1.0.11.

**Comment by ethought** [comment]
@daira Thank you, will wait for 1.0.11

**Comment by ioptio** [comment]
@ethought can you confirm your issue was fixed in 1.0.11?
